proxy_worker/dispatcher.py

```diff
@@ -0,0 +1,521 @@
+        disp = cls.__current_dispatcher__
+        if disp is None:
+            raise RuntimeError('no currently running Dispatcher is found')
+        return disp
+
+
+class Dispatcher(metaclass=DispatcherMeta):
+    _GRPC_STOP_RESPONSE = object()
+
+    def __init__(self, loop: AbstractEventLoop, host: str, port: int,
+                 worker_id: str, request_id: str,
+                 grpc_connect_timeout: float,
+                 grpc_max_msg_len: int = -1) -> None:
+        self._loop = loop
+        self._host = host
+        self._port = port
+        self._request_id = request_id
+        self._worker_id = worker_id
+        self._grpc_connect_timeout: float = grpc_connect_timeout
+        self._grpc_max_msg_len: int = grpc_max_msg_len
+        self._old_task_factory: Optional[Any] = None
+
+        self._grpc_resp_queue: queue.Queue = queue.Queue()
+        self._grpc_connected_fut = loop.create_future()
+        self._grpc_thread: Optional[threading.Thread] = threading.Thread(
+            name='grpc_local-thread', target=self.__poll_grpc)
+
+        self._sync_call_tp: Optional[concurrent.futures.Executor] = (
+            self._create_sync_call_tp(self._get_sync_tp_max_workers()))
+
+    def on_logging(self, record: logging.LogRecord,
+                   formatted_msg: str) -> None:
+        if record.levelno >= logging.CRITICAL:
+            log_level = protos.RpcLog.Critical
+        elif record.levelno >= logging.ERROR:
+            log_level = protos.RpcLog.Error
+        elif record.levelno >= logging.WARNING:
+            log_level = protos.RpcLog.Warning
+        elif record.levelno >= logging.INFO:
+            log_level = protos.RpcLog.Information
+        elif record.levelno >= logging.DEBUG:
+            log_level = protos.RpcLog.Debug
+        else:
+            log_level = getattr(protos.RpcLog, 'None')
+
+        if is_system_log_category(record.name):
+            log_category = protos.RpcLog.RpcLogCategory.Value('System')
+        else:  # customers using logging will yield 'root' in record.name
+            log_category = protos.RpcLog.RpcLogCategory.Value('User')
+
+        log = dict(
+            level=log_level,
+            message=formatted_msg,
+            category=record.name,
+            log_category=log_category
+        )
+
+        invocation_id = get_current_invocation_id()
+        if invocation_id is not None:
+            log['invocation_id'] = invocation_id
+
+        self._grpc_resp_queue.put_nowait(
+            protos.StreamingMessage(
+                request_id=self.request_id,
+                rpc_log=protos.RpcLog(**log)))
+
+    @property
+    def request_id(self) -> str:
+        return self._request_id
+
+    @property
+    def worker_id(self) -> str:
+        return self._worker_id
+
+    @classmethod
+    async def connect(cls, host: str, port: int, worker_id: str,
+                      request_id: str, connect_timeout: float):
+        loop = asyncio.events.get_event_loop()
```

Comment:
Add a task - app setting for enabling "uvloop" - add this feature.

`uvloop.new_event_loop` vs `asyncio.events.get_event_loop()`